sound/soc/sof/sof-audio.c

```diff
@@ -380,49 +401,34 @@ int sof_widget_list_setup(struct snd_sof_dev *sdev, struct snd_sof_pcm *spcm, in
 			dev_err(sdev->dev, "error: no pipeline widget found for %s\n",
 				swidget->widget->name);
 			ret = -EINVAL;
-			goto pipe_free;
+			goto widget_free;
 		}
 
-		ret = sof_widget_setup(sdev, pipe_widget);
-		if (ret < 0)
-			goto pipe_free;
-
-		num_pipe_widgets++;
-
 		if (pipe_widget->complete)
 			continue;
 
 		pipe_widget->complete = snd_sof_complete_pipeline(sdev->dev, pipe_widget);
 		if (pipe_widget->complete < 0) {
 			ret = pipe_widget->complete;
-			goto pipe_free;
+			goto widget_free;
 		}
 	}
 
 	return 0;
 
-pipe_free:
-	/* free the pipe widgets that were set up successfully */
+widget_free:
+	/* free all widgets that have been set up successfully */
 	for_each_dapm_widgets(list, i, widget) {
 		struct snd_sof_widget *swidget = widget->dobj.private;
 
 		if (!swidget)
 			continue;
 
-		if (!num_pipe_widgets--)
-			break;
-
-		sof_widget_free(sdev, swidget->pipe_widget);
-	}
-
-widget_free:
-	/* free all widgets that have been set up successfully */
-	for_each_dapm_widgets(list, i, widget) {
 		if (!num_widgets--)
 			break;
 
-		if (widget->dobj.private)
-			sof_widget_free(sdev, widget->dobj.private);
+		sof_widget_free(sdev, swidget);
+		sof_widget_free(sdev, swidget->pipe_widget);
```

Comment:
I am still not completely clear on this.

A pipeline is a widget. So when you setup an pipeline, do you count it twice, and likewise here would you reduce the refcount twice?